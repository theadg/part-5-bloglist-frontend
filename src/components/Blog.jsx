import { useState } from 'react'

const Blog = ({ blog, likeBlog, deleteBlog, user }) => {
    const [showDetails, setShowDetails] = useState(false)

    const blogStyle = {
        paddingTop: 10,
        paddingLeft: 2,
        border: 'solid',
        borderWidth: 1,
        marginBottom: '1rem',
    }

    return (
        <div className="blog-item" style={blogStyle}>
            <div data-testid="blogTitle">{blog.title}</div> by
            <div data-testid="blogAuthor">{blog.user?.name ?? 'No Author'}</div>
            <button
                data-testid="showButton"
                type="button"
                onClick={() => setShowDetails(!showDetails)}>
                {!showDetails ? 'View' : 'Hide'}
            </button>
            <br />
            {showDetails && (
                <div className="">
                    <div data-testid="blogUrl">{blog.url}</div>
                    <div data-testid="blogLikes">Likes: {blog.likes}</div>
                    <br />
                    <button
                        type="button"
                        data-testid="blogLikeBtn"
                        onClick={() => likeBlog(blog)}>
                        like
                    </button>
                    <br />
                    {blog?.user?.name === user?.name && (
                        <button
                            type="button"
                            data-testid="deleteBtn"
                            onClick={() => deleteBlog(blog)}>
                            delete
                        </button>
                    )}
                </div>
            )}
        </div>
    )
}

export default Blog
