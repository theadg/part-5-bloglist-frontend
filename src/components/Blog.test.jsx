import React from 'react'
import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import '@testing-library/jest-dom'
import Blog from './Blog'
import renderer from 'react-test-renderer'

test('renders content', () => {
    const blog = {
        title: 'sample',
        user: { name: 'badg' },
        url: 'hello.com',
        likes: 0,
    }

    render(<Blog blog={blog} />)

    const title = screen.getByTestId('blogTitle')
    const author = screen.getByTestId('blogAuthor')
    const url = screen.queryByTestId('blogUrl')
    const likes = screen.queryByTestId('blogLikes')

    expect(title).toHaveTextContent(blog.title)
    expect(author).toHaveTextContent(blog.user.name)
    expect(url).toBeNull()
    expect(likes).toBeNull()
})

test('renders content (snapshot)', () => {
    const blog = {
        title: 'sample',
        user: { name: 'badg' },
        url: 'hello.com',
        likes: 0,
    }

    const tree = renderer.create(<Blog blog={blog} />).toJSON()

    expect(tree).toMatchSnapshot()
})

test('shows url and likes on click', async () => {
    const blog = {
        title: 'sample',
        user: { name: 'badg' },
        url: 'hello.com',
        likes: 0,
    }

    render(<Blog blog={blog} />)

    const showBtn = screen.getByTestId('showButton')
    const user = userEvent.setup()
    await user.click(showBtn)

    const url = screen.getByTestId('blogUrl')
    const likes = screen.queryByTestId('blogLikes')

    expect(url).toHaveTextContent(blog.url)
    expect(likes).toHaveTextContent(blog.likes)
})

test('like button is clicked twice and like fn is called twice', async () => {
    const blog = {
        title: 'sample',
        user: { name: 'badg' },
        url: 'hello.com',
        likes: 0,
    }

    const mockHandler = jest.fn()

    render(
        <Blog
            blog={blog}
            likeBlog={mockHandler}
        />
    )

    const user = userEvent.setup()
    const showBtn = screen.getByTestId('showButton')
    await user.click(showBtn)

    const likeBtn = screen.getByTestId('blogLikeBtn')
    await user.dblClick(likeBtn)

    expect(mockHandler.mock.calls).toHaveLength(2)
    screen.debug(showBtn)
})
