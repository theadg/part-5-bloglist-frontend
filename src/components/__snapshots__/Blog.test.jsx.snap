// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders content (snapshot) 1`] = `
<div
  style={
    {
      "border": "solid",
      "borderWidth": 1,
      "marginBottom": "1rem",
      "paddingLeft": 2,
      "paddingTop": 10,
    }
  }
>
  <div
    data-testid="blogTitle"
  >
    sample
  </div>
   by
  <div
    data-testid="blogAuthor"
  >
    badg
  </div>
  <button
    data-testid="showButton"
    onClick={[Function]}
    type="button"
  >
    View
  </button>
  <br />
</div>
`;
